Edit of "Output-function summary" docs from a new learner (#56767)
JuliaLang/julia: doc/src/manual/types.md

### Output-function summary

Here is a brief summary of the different output functions in Julia and how they are related.
Most new types should only need to define `show` methods, if anything,
since the other functions mentioned here (except `write`) call `show`
in the absence of a more specific method.

* [`show(io, x)`](@ref), with two arguments,
  is the default simple text representation of `x`.
  It is typically the format you would employ to input `x` into Julia.

> * [`show(io, mime, x)`](@ref), with three arguments,

Reviewer: ```suggestion
* [`show(io, ::MIME"text/plain", x)`](@ref), with three arguments,
```

Author: I do not understand why you want this bullet to be specific to a text representation rather than explaining the 3-argument `show` generally.

Reviewer: Because text/plain is the common case, and most of this section is about text output . Most types don’t offer output for any other mime types, which is why they are only mentioned in passing here. 

Also, if you want to overload it you need to know the specific signature.

  performs verbose pretty-printing of `x`.

>   Multiple 3-argument `show` methods can be defined for various MIME types to enable

Reviewer: ```suggestion
  Additional 3-argument `show` methods can be defined for other MIME types to enable
```

Author: This change is contingent on the decision above: `mime` vs `::MIME"text/plain"`.

>   richer display of `x` in some interactive environments as discussed above.

Reviewer: ```suggestion
  richer display of `x` in some interactive environments as discussed below.
```

Author: Hmm, when I read this passage I thought "above" referred to the `Polar` example in the previous section. Now I think you may have been referring to the small explanation in the `display` bullet. I will rethink this.

>   By default (if no 3-argument method is defined for `typeof(x)`),

Reviewer: ```suggestion
  Although the underlying method is defined for a [`MIME`](@ref) argument, the caller can
  simply pass a string such as `"text/plain"`, at the expense of a dynamic dispatch.
  By default (if no 3-argument method is defined for `typeof(x)`),
```

Author: I think this is good to explain, but it leaves me with more questions than answers. Why is the string version used most places if it is slower? How is `::MIME"text/plain"` related to `MIME"text/plain"`? (I have seen definitions for both.)

Reviewer: Internally, we should call the MIME version.  For users, the string version is easier, and is rarely performance-critical.

>   it calls the 2-argument `show(io, x)`.

Reviewer: ```suggestion
  `show(io, "text/plain", x)` calls the 2-argument `show(io, x)`.
```

Author: This sentence is copied from the original text, but I am thinking now it can just be deleted? 

If not deleted, then maybe moved into the `display` bullet like this?
> In the REPL, `display` calls `show(io, MIME"text/plain", x)` (which defaults to `show(io, x)` if not separately defined).

Reviewer: It’s copied from the previous text, but you edited the preceding sentence so that the new qualification is needed.

* [`print(io, x)`](@ref) by default calls `show(io, x)`,
  but a few types have a distinct `print` format —
  most notably, when `x` is a string, `print` outputs the raw text
  whereas `show` outputs an escaped string enclosed in quotation marks.
  ([`println`](@ref) and [`printstyled`](@ref) both call [`print`](@ref),
  but add a newline or styling, respectively.)

* [`display(x)`](@ref) tells the current environment to display `x`
  in whatever way it thinks best.
  This is the function used by the REPL to output the result of an evaluated expression.
  In the REPL, `display` calls `show(io, MIME"text/plain", x)`.

>   In a notebook, like Jupyter or Pluto, `display` calls `show(io, MIME"text/html", x)`
>   (or sometimes `show(io, MIME"image/<format>", x)` if `x` is representable as an image).

Reviewer: ```suggestion
  In a graphical environment, such as Jupyter or Pluto, `display` might prefer a non-plaintext
  representation of an object (such as HTML, Markdown, or a PNG or SVG image), if the
  corresponding 3-argument `show` method is defined.
```

* [`write(io, x)`](@ref), if it is defined
  (it generally has *no* default definition for new types),
  writes a "raw" binary representation of `x` to `io`,
  e.g. an `x::Int32` will be written as 4 bytes.

> The first argument to these functions, `io`, defines where the output should be written.
> In the REPL, `io` is an [`IOContext`](@ref) wrapper around [`stdout`](@ref).
> [`stdout`](@ref) directs output to the default display, typically the terminal window.
> The `io` argument for all the above functions defaults to `stdout` if it is omitted.

Reviewer: ```suggestion
The `io` argument for all the above functions is optional and defaults to [`stdout`](@ref) if it is omitted;
this is the default output stream, typically a terminal window.

More generally, `io` (of type [`IO`](@ref)) specifies a desired output stream (such as a file, buffer, or pipe).
In the REPL, `io` is an [`IOContext`](@ref) wrapper around `stdout`, as described below.
```

It is also helpful to be familiar with the metadata that can be attached to an `io` stream
by an [`IOContext`](@ref) wrapper.
For example,
the REPL sets the `:limit => true` flag from `display` for an evaluated expression,
in order to limit the output to fit in the terminal;
you can query this flag with `get(io, :limit, false)`.
And when displaying an object contained within, for example, a multi-column matrix,
the `:compact => true` flag could be set,
which you can query with `get(io, :compact, false)`.

The functions `string`, `annotatedstring`, and `repr`
may be used to save values to a string (rather than output to `io`).
These functions call either `show` or `print` as indicated below,
so it is usually not necessary to define new methods for them separately.
Note that [`convert(String, x)`](@ref) is purposely not defined

> so the user can be explicit about the string representation they desire.

Reviewer: I think that the real reason is that `convert` in Julia is used for *implicit* conversions, e.g. when assigning an expression to an container of a different type.  So it is *never* defined for converting between conceptually different types of objects, e.g. between numbers and strings.

Author: I have now rewritten this sentence according to your explanation.

In the REPL, `io` is an [`IOContext`](@ref) wrapper around `stdout`, as described below.
It is also helpful to be familiar with the metadata that can be attached to an `io` stream
by an [`IOContext`](@ref) wrapper.
For example,
the REPL sets the `:limit => true` flag from `display` for an evaluated expression,
in order to limit the output to fit in the terminal;
you can query this flag with `get(io, :limit, false)`.
And when displaying an object contained within, for example, a multi-column matrix,
the `:compact => true` flag could be set,
which you can query with `get(io, :compact, false)`.

The functions `string`, `annotatedstring`, and `repr`
may be used to save values as a string (rather than output to `io`).
These functions call either `show` or `print` as indicated below,
so it is usually not necessary to define new methods for them separately.
(Note that [`convert(String, x)`](@ref) is not defined for most `x`
since the `convert` function is intended for implicit conversions.)

* [`string(x)`](@ref) and [`annotatedstring(x)`](@ref) call [`print(io, x)`](@ref).
* [`repr(x)`](@ref), with one argument, calls the 2-argument [`show(io, x)`](@ref).

> * `repr("text/plain", x)`, with two arguments, calls the 3-argument [`show(io, ::MIME"text/plain", x)`](@ref).

Author: Can this be generalized to `repr(mime, x)` $\rightarrow$ `show(io, mime, x)`?
I also do not understand the change of form from `"text/plain"` to `::MIME"text/plain"`. Does either option work in either location?

Reviewer: Yes, it can be generalized I think. 

(Writing from my phone so I can’t double check)

### Output-function summary

Here is a brief summary of the different output functions in Julia and how they are related.
Most new types should only need to define `show` methods, if anything,
since the other functions mentioned here (except `write`) call `show`
in the absence of a more specific method.

* [`show(io, x)`](@ref), with two arguments,
  is the default simple text representation of `x`.
  New types will automatically define `show(io, x)`
  as the format employed to input `x` into Julia.
  However, it may be desirable to overwrite this default

>   to define nicer printing for collections of the new type.

Author: I'm not totally confident on why when/why to overwrite 2-arg show. Printing inside a vector is the example used above for `Polar`, but the `show` docstring maintains that the definition "should be parseable Julia code when possible" which seems to already be the default definition.
